### ISSUE TITLE
Are generic classes type-checked?

### ISSUE THREAD

**Issue by tslettebo** [facebook/hhvm#5110]
Using the example at this page (http://docs.hhvm.com/manual/en/hack.generics.class.php) and adding the following lines, I had expected the type-checker to complain, but it doesn't:

$mail = new MailBox();

$mail->put(1234);
$mail->put("Test");

The example code says:

// The mailbox can contain any type, but, per instantiation, once associated
// with a type, it cannot change. Mailbox<int>, Mailbox<string>, Mailbox<mixed>

I'd expect that in the first put() call, it would associate T with int, so the next put() call would fail the type-checker, but it doesn't.

**Comment by paulbiss** [comment]
The problem here is that the first call to `put` doesn't actually resolve the type.
```
$mail = new MailBox();
// $mail has type MailBox<unresolved>
$mail->put(1234);
// $mail has type MailBox<unresolved of (int)>
$mail->put("Test");
// $mail has type MailBox<unresolved of (int, string)>
```

Suppose you have a function `function foo(MailBox<int>): void`, then,
```
$mail = new MailBox();
// $mail has type MailBox<unresolved>
$mail->put(1234);
// $mail has type MailBox<unresolved of (int)>
foo($mail);
// $mail has type MailBox<int>
$mail->put("Test"); // <- error here
```

OTOH if you had a another function `function bar(MailBox<mixed>): void`, then,
```
$mail = new MailBox();
// $mail has type MailBox<unresolved>
$mail->put(1234);
// $mail has type MailBox<unresolved of (int)>
bar($mail);
// $mail has type MailBox<mixed>
$mail->put("Test"); // no error as string is compatible with MailBox<mixed>
```

@jwatzman can probably give a better explanation.

**Comment by tslettebo** [comment]
Thanks for a quick reply. I realised afterwards, reading this page (http://docs.hhvm.com/manual/en/hack.generics.typeinference.php) that this is the case, and I was going to remove my question or add a note to  it, but then I saw your reply.

**Comment by jwatzman** [comment]
For the record, yeah @paulbiss has the right explanation. If you want really nitty gritty details on the type system, I gave a talk about it a few weeks ago [which is on YouTube](https://www.youtube.com/watch?v=aN22-V-b8RM).

**Comment by tslettebo** [comment]
Thanks, I'll check it up. :)

By the way, great work, all of you at Facebook, for HHVM/Hack!

I've been working with PHP for 12 years, and before that, I've been working with C++ and Java. I've been really missing the ability to be more explicit in the code, to be able to put more of my *intentions* in the code, and having it checked by the compiler or runtime.

For years, I've been looking for a possible alternative to PHP, either a statically typed version of it, or another language, although I knew that, as we have relatively large systems, a complete rewrite in another language would probably not be feasible, economically.

Then along came Hack, and my search was over. :)

These days, I'm testing it out, and learning how to get our code type-checked (going straight for "strict" where possible).

One of the challenges is how to go from "untyped data" (such as web form submissions and database recordsets) to typed data, and the "Hack cookbooks" has helped me understand this better.

The idea of "lazy typing" (use of "unresolved" that gets gradually better defined) is completely new to me, as I guess it is to many others. Lazy evaluation (like in Haskell) is kind of more familiar. :)

Also, in other statically typed languages, there's no such thing as "mixed" (maybe the closest is "void *" in C++ and "Object" in Java), so I've been wondering how to do *anything* with something of type "mixed", since I always got error messages from the type checker if I assumed a certain type.

Now I've realised that we need to make that assumption *explicit* in the code, by testing for certain types, and when we do, the type checker knows that it has a more specific type.

I realise that having code like if($GET["something"] == 1234) ... reveals a hidden assumption - that this array element is of type (convertible to) int - and that even if *I* know that the data is of the right type, the type checker doesn't know that, and can't check it for me.

I've been wondering how to go from untyped to typed data in a general manner, without having to write a function for every kind of type there is. I've been thinking of using generics for it, but as it seems to be based on type erasure (like Java's generics) it seems it's not possible.

Like, I can't do something like this (free function):

function get<T>(array<string,mixed> $map, string $key) : T
{
  if(gettype($map[$key]) == T)
    return $map[$key];
  else
    throw new Exception("Type mismatch");
}

I guess this doesn't work because T isn't an actual "type variable", unlike C++ templates, for example.

I know this isn't the right forum (and I've gone way beyond the original subject line of this thread already), but another thing that would be nice is operator overloading.

We have a Money class in our system, and using dimensional analysis, we ensure that we get the right unit at the end, e.g. PricePerMm * Mm = Price (Money) (this is a newspaper ad system).

However, since there's no operator overloading, we have to write things like: $price_per_mm->multiply($mm), which becomes quite unreadable for more complex expressions.

I understand operator overloading is already used for the GMP extension, so it obviously can be done, so why can't it be made available also for "userland" code?

**Comment by tslettebo** [comment]
| @jwatzman For the record, yeah @paulbiss has the right explanation. If you want really nitty gritty details on the type system (including this feature and a lot more), I gave a talk about it a few weeks ago which is on YouTube.

Thanks for the link, I've watched it now, and it helped me understand better how the type system works.

I was also at some point (in some forum, probably the Hack Facebook group would be the best place) going to ask if you guys have considered adding support for "const", and I see you responded to a question about that in the presentation.

For me, the big thing about "const" isn't necessarily optimisation possibilities, but rather about improving our ability to *reason* about the code, and as I notice you're aware of, this is a big thing in languages with support for immutability in various forms, such as C++'s "const" or immutability in functional languages.

In other words, the value of immutability may be not at least in optimising *productivity*, rather than performance.

I'm happy that this is under consideration, and I see some signs of it, such as the immutable collections.
